### PR TITLE
JavaScript Capstone Project

### PR DESCRIPTION
**Pull request by ishimwepati**
Our JavaScript Capstone project was a collaborative effort between myself, @Xghayor, and my partner, @ishimwepati. Together, we created a captivating web application that revolves around showcasing restaurant meals in an engaging way.

Here's a rundown of what we accomplished:

### Using APIs
We integrated two important APIs into our project:
1. **Meal DB API**: This allowed us to gather a list of unique items along with detailed information about each meal. This formed the foundation of our meal display.
2. **Involvement API**: We used this API to record user interactions like likes and comments.

### Building Interfaces
Our project featured two user interfaces designed to offer an enjoyable experience:
1. **Home Page**: The central hub of our app. When users land on this page, they see a variety of meals retrieved from the Meal DB API. They can like items, view comments, and even engage in discussions.
2. **Comments Popup**: This elegant popup displayed additional details about selected meals and showcased comments from users. It provided an avenue for users to contribute their thoughts and see real-time updates.

### Special Features
- **Design and Navigation**: We paid careful attention to creating a user-friendly interface, sticking to the original wireframes while adding our own unique design elements through colors, fonts, and spacing.
- **Counters**: Throughout the interfaces, we included counters that dynamically show the number of items, comments, and reservations (for specific scenarios). These numbers updated automatically as users interacted with the app.
- **Interaction Tracking**: When users liked an item or left a comment, our app recorded these interactions using the Involvement API, ensuring that the interface stayed up-to-date.
- **Reservations Popup**: For specific cases, we added a popup that mimicked the functionality of the Comments Popup but for making reservations.

### Technical Details
To ensure a smooth development process, we followed these steps:
- **GitHub and Gitflow**: We used GitHub with the Gitflow approach to effectively manage our collaboration and version control.
- **Webpack**: We employed Webpack to bundle our project, which made development modular and improved performance.
- **Jest Testing**: For quality assurance, we used Jest, a JavaScript testing library, to write unit tests for our count functions and validate our code.

In essence, our project offers a delightful journey into the world of restaurant meals, inviting users to explore, engage, and share their experiences. It blends technology and design to create an interactive platform that celebrates the pleasures of good food and community connections.

Your suggestions are most welcome !

### CODE REVIEW

File: src/modules/displayMeals.js [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+      return [];
+    }
+
+    const data = await response.json();
+    return data;
+  } catch (error) {
+    console.error('Error fetching comments:', error);
+    throw error;
+  }
+};
```
Comment (AbelG101) [inline src/modules/displayMeals.js:103]: - [ ] Ouch! you're doing so well however as part of the project requirements the functionality for the comments counter should be implemented as a separate module, so I'd advise you to create a separate module for fetching the number of comments as this should be covered in the unit tests.

File: src/modules/homeCounter.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const homeCounter = (length) => {
+  const homeIcon = document.getElementById('home-icon');
+
+  const counterSpan = document.createElement('span');
+  counterSpan.textContent = `(${length})`;
+
+  homeIcon.appendChild(counterSpan);
+};
+
+export default homeCounter;
```
Comment (AbelG101) [inline src/modules/homeCounter.js:10]: - [ ] Please make sure your counter functions make the counting based on what is actually displayed on the page instead of relying on the amount of data you get from the API. Also, make sure to cover all the edge cases you can think about.
- [ ] This applies to the comments counter you'll create as well. ✔

File: src/modules/tests/displayMeals.test.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  test('does not update comment counter if element is not found', () => {
+    document.body.innerHTML = '';
+
+    updateCommentCounter(10);
+
+    const commentCounterElement = screen.queryByText('Comments: 10');
+    expect(commentCounterElement).toBeNull();
+  });
+});
```
Comment (AbelG101) [inline src/modules/tests/displayMeals.test.js:25]: - [ ] Kindly make sure that your tests for the counter functions focus on testing how many items are on the page by invoking the counter function and also make sure to cover all edge cases such as when there are 0 items on the page (counter function should return 0). This applies to the unit test for the comments counter as well.

**Review by AbelG101 — CHANGES_REQUESTED** [review]
# Hi @ishimwepati & @Xghayor 👋, 

You're doing a good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- The web app properly implements API requests ✔
- Data can be properly fetched and updated ✔
- All pages are properly displayed ✔
- Proper use of ES6 syntax ✔
- Correct use of Git Flow ✔
-  Well-structured files ✔
- All linter checks are passing ✔
- Professionally documented readme ✔
- Descriptive PR title with a top-notch summary ✔

## Required Changes ♻️
- _Check the comments under the review._
- you just have to make a few changes and you're on your way 🚀.


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/modules/displayMeals.js (hunk 1 above)
Comment (ishimwepati) [reply src/modules/displayMeals.js:104]: Thanks a lot !

Comment (Abenezer-Tilahun) [reply src/modules/displayMeals.js:97]: Fixed.

File: src/modules/homeCounter.js (hunk 2 above)
Comment (Abenezer-Tilahun) [reply src/modules/homeCounter.js]: Fixed.

File: src/modules/tests/displayMeals.test.js (hunk 3 above)
Comment (Abenezer-Tilahun) [reply src/modules/tests/displayMeals.test.js]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @ishimwepati @Xghayor 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.

## Optional suggestions 
 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._
